src/EditorFeatures/Core/BraceMatching/AbstractEmbeddedLanguageBraceMatcher.cs

```diff
@@ -2,44 +2,154 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Editor;
-using Microsoft.CodeAnalysis.Editor.EmbeddedLanguages;
 using Microsoft.CodeAnalysis.EmbeddedLanguages;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
 using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.BraceMatching
 {
-    // Note: this type could be concrete, but we cannot export IBraceMatcher's for multiple
-    // languages at once.  So all logic is contained here.  The derived types only exist for
-    // exporting purposes.
-    internal abstract class AbstractEmbeddedLanguageBraceMatcher : IBraceMatcher
+    /// <summary>
+    /// Brace matcher that analyzes string literals (for C#/VB) and then dispatches out to embedded brace matchers for
+    /// particular embedded languages (like JSON/Regex).
+    /// </summary>
+    internal abstract class AbstractEmbeddedLanguageBraceMatcher :
+        AbstractEmbeddedLanguageFeatureService<IEmbeddedLanguageBraceMatcher>, IBraceMatcher
     {
+        protected AbstractEmbeddedLanguageBraceMatcher(
+            string languageName,
+            EmbeddedLanguageInfo info,
+            ISyntaxKinds syntaxKinds,
+            IEnumerable<Lazy<IEmbeddedLanguageBraceMatcher, EmbeddedLanguageMetadata>> allServices)
+            : base(languageName, info, syntaxKinds, allServices)
+        {
+        }
+
         public async Task<BraceMatchingResult?> FindBracesAsync(
             Document document, int position, BraceMatchingOptions options, CancellationToken cancellationToken)
         {
-            var languagesProvider = document.GetLanguageService<IEmbeddedLanguagesProvider>();
-            if (languagesProvider != null)
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            return FindBraces(semanticModel, position, options, cancellationToken);
+        }
+
+        private BraceMatchingResult? FindBraces(
+            SemanticModel semanticModel,
+            int position,
+            BraceMatchingOptions options,
+            CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<IEmbeddedLanguageBraceMatcher>.GetInstance(out var serviceBuffer);
+            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
+            var worker = new Worker(this, semanticModel, position, options, serviceBuffer, cancellationToken);
+            return worker.Recurse(root);
```

Comment:
we have a position - why do we need to explore starting at the root?  is it not possible to find the node at the position, see if its in a string and then dispatch?

Reply:
yes.  that's a fair point.  i'll just switch to a direct lookup of the token.